### ISSUE TITLE
Huge number of "Forwarding request for shard" and "Resolve of path sequence XXX failed" messages

### ISSUE THREAD

**Issue by object** [akkadotnet/akka.net#3522]
I am trying to make sense of our debug logs flooded with these messages. I checked our system activities for 3 hours and here are some data:

Within 3 hours:
- created 240 aggregate roots (actors that use cluster sharding)
- these roots received not more than 4000 messages 
the log contains:
- "Forwarding request for shard" 11.6M (11.6 million entries, ca. 1000 entries per second!) - this line is from Akka.NET code (ShardRegions.cs - DeliveryMessage method)
- "Resolve of path sequence XXX failed" 1.6M - they didn't appear in logs for the first 30 min. after restart of services - this is also from Akka.NET code (ActorRefProvider.cs - method ResolveActorRef)

The system more or less worked - it handled business logic properly most of that time.

If I stopped all nodes but one, i.e. there will be only a single node in a role that uses cluster sharding, then the flood of these log entries stops.

**Comment by Aaronontheweb** [comment]
@object got a theory here, since in Gitter chat you mentioned that this occurs when `remember-entities` is on - do you know if these logs occur after / around a re-balancing operation of any kind?

**Comment by object** [comment]
@Aaronontheweb they might though nothing indicates they did. Both our Stage and Test environments had 2 nodes in that role when this happened, with split brain resolver (keep majority) and remember-entities turned on. During the observation time there were no deployments on these servers, neither I can see that services were restarted, so in principle there shouldn't be relocations of shards. But the fact that these crazy activities started with some (30 min.) delay after start of services may indicate that there was some rebalancing. I will dig into the logs around the time it all started.

**Comment by object** [comment]
Hmm, I also see that message "Received gossip status" is only written in logs in our Test environment. So we _might_ have different log filtering in Test and Stage.

**Comment by Aaronontheweb** [comment]
@object my guess is that you're running with `DEBUG` turned on in test but not in staging. This issue probably occurs in both environments, but you probably run with an `INFO` setting in Stage which is what most production users do.

**Comment by Aaronontheweb** [comment]
> NB! I was wrong. Logging activities started right after the deployment, so it was not due rebalancing but on cluster restart.

That's a great piece of information - thank you. Narrows down the search area big time.

**Comment by object** [comment]
@Aaronontheweb Must be so. So let's forget about our Stage environment, it's Test where we reproduce the problem.

**Comment by object** [comment]
I will edit the issue to remove information about different environments, it's misleading.

**Comment by Aaronontheweb** [comment]
> If I stopped all nodes but one, i.e. there will be only a single node in a role that uses cluster sharding, then the flood of these log entries stops.

A single node that hosts shards? Or a single node that communicates with the shards hosted elsewhere?

I.E. are these logs coming from a shard region or a shard region proxy, if you're using a mix of both?

**Comment by object** [comment]
Single node that hosts shards. The logs are coming from a shard region.

**Comment by pmbanka** [comment]
I found the root cause of this particular problem. It was a bug in our message extractor - it was routing messages based on their hash code. It was doing this both for our messages AND Akka messages as well, like `ShardRegion.StartEntity`. Once I added special handling for `ShardRegion.StartEntity`, the problem disappeared.

There is also a question I have about the Akka messages - the code example in the documentation https://getakka.net/articles/clustering/cluster-sharding.html#remembering-entities

```
public sealed class MessageExtractor : HashCodeMessageExtractor
{
    public MessageExtractor() : base(maxNumberOfShards: 100) { }

    public string EntityId(object message) 
    {
        switch(message)
        {
            case ShardEnvelope e: return e.EntityId;
            case ShardRegion.StartEntity start: return start.EntityId;
        }
    } 
    public object EntityMessage(object message) => (message as ShardEnvelope)?.Message ?? message;
}
```

doesn't seem to be compilable, because not all code paths return a value. So, my question is - how should other `ShardRegion` messages be handled? I can extract the data from `StartEntityAck`, but some other of these messages are internal, and I can't do anything about it. What should be set as `entityId` for them?

**Comment by zbynek001** [comment]
in case the `message ` is `StartEntity` or your shard message, you should return `EntityId`.
In all other cases you should return `null`

**Comment by pmbanka** [comment]
@zbynek001 I may be wrong, but it looks like returning null will result with NRE (at least when used with `HashCodeMessageExtractor`). 

Here, `EntityId` will assing null to `id`
https://github.com/akkadotnet/akka.net/blob/41c116d2a3d280481e584401608ec848dc147632/src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs#L98

and here it will explode 
https://github.com/akkadotnet/akka.net/blob/41c116d2a3d280481e584401608ec848dc147632/src/core/Akka/Util/MurmurHash.cs#L205

**Comment by zbynek001** [comment]
What I send above was about message extractors in general, not about specifics like `HashCodeMessageExtractor`

But anyway, in case of `HashCodeMessageExtractor`

- ShardRegion will try to extract `EntityId`
- Only when it's not null it'll try to get `ShardId`

so the code you referenced should not be executed in that scenario

**Comment by pmbanka** [comment]
Ah, that makes sense! Thanks for the clarification :)

**Comment by Aaronontheweb** [comment]
So, can I close this?

**Comment by object** [comment]
Let's wait for status from @pmbanka. We had some problems today, but I guess he should have an update on the issue soon.

**Comment by pmbanka** [comment]
@Aaronontheweb I digged into it today and I realised that 
1. There was indeed an issue in our extractor
2. BUT fixing the issue did not make the problem go away - I was wrong with that point.

I managed to create a reasonably small repro that shows the issue we encounter - the code is here https://github.com/pmbanka/akka-shards-spam-repro

Repro steps and later findings are all in Readme.txt.

**Comment by zbynek001** [comment]
you have there an issue with `ShardId` extracting.
It should also process `ShardRegion.StartEntity` messages and transform `EntityId` same way as from your objects.
This is required for remember-entities feature.

**Comment by zbynek001** [comment]
actually, looking at it again, there are more potential issues.
First, to clarify something:

what is your expected EntityId?
looking at the code, there will be only one entity "the_one_and_only". Meaning all messages will go to one actor with `EntityId` = "the_one_and_only".
But then the `ShardId` will be wrong, because it'll return different values for the same `EntityId`. That's not supported

**Comment by pmbanka** [comment]
@zbynek001 It is quite possible that I don't understand something fundamental here :)

I was under the impression that an `entityId` does not need to be globally unique per cluster, and it only needs to be unique per particular shard. I thought a pair of `shardId * entityId` has to be globally unique in order to identify an actor. Is this assumption wrong?

The first extractor produces pairs of shardId/entityId like that:

/1/the_one_and_only
/2/the_one_and_only

so if my assumption is wrong, then yeah - this will simply not work.

But! there is also another extractor causing the same issues. This one produces pairs of shardId/entityId like that:

/1/1
/2/2

in this setup, the entityIds are in fact unique cluster-wise.

Moreover, there is also a weirdness I observed that the system works if the entities has been already spawned before second node joined the cluster, so there still may be something more to that issue...

BTW, thanks for looking into that :) I have to leave now but I'll try to crawl through the Akka codebase tomorrow to see how exactly are these EntityId/ShardId values used.

**Comment by object** [comment]
Reply from @Horusiath: "in general entities are globally identified by shardId * entityId pair, but if remember-entities option is used, you must be able to infer shardId based on entityId alone (don't ask me why, lightbend team decided to do it this way)"

**Comment by pmbanka** [comment]
Hah, so I looked into it again today and I realised the problem is indeed with our extractor code, not with Akka :)

Our extractor uses GetHashCode, which I just learned [produces different results for same strings across different AppDomains (so, different runs of application as well)](https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/). Akka's implementation of `HashCodeMessageExtractor` uses `Akka.Util.MurmurHash`, which produces deterministic hashes. Using that in my repro fixes the problem.

So, TL;DR - I didn't know how GetHashCode works, issue can be closed :)

**Comment by object** [comment]
Wow! Well done, Pawel. Actually I knew that GetHashCode is not deterministic across domains (there was a Twitter thread about it a year ago or so) but wouldn't that think about it in the context Akka clustering. I am pretty sure we are not only who hit this problem.

**Comment by Aaronontheweb** [comment]
Glad you worked that out. Made this mistake myself before in my own applications - it's one of those "gotchas" that I think you can only really learn by being burned by it once. The hash code randomization serves a purpose (security - prevents people from being able to predict the hash of a specific value in advance for any given .NET process) but that lack of consistency makes it useless for distributed computing.

Using our `MurmurHash` implementation is the way to go.
